tool/go

```diff
@@ -0,0 +1,103 @@
+#!/bin/sh
+#
+# This script acts like the "go" command, but uses Tailscale's
+# currently-desired version from https://github.com/tailscale/go,
+# downloading it first if necessary.
+
+set -eu
+
+log() {
+    echo "$@" >&2
+}
+
+DEFAULT_TOOLCHAIN_DIR="${HOME}/.cache/tailscale-go"
+TOOLCHAIN="${TOOLCHAIN-${DEFAULT_TOOLCHAIN_DIR}}"
+TOOLCHAIN_GO="${TOOLCHAIN}/bin/go"
+read -r REV < "$(dirname "$0")/../go.toolchain.rev"
+
+# Fast, quiet path, when Tailscale is already current.
+if [ -e "${TOOLCHAIN_GO}" ]; then
+    short_hash=$("${TOOLCHAIN_GO}" version | sed 's/.*-ts//; s/ .*//')
+    case $REV in
+    "$short_hash"*)
+        unset GOROOT
+        exec "${TOOLCHAIN_GO}" "$@"
+    esac
+fi
+
+set +u
```

Comment:
HACK: You have to `set +u` (disable exploding on undefined envvars) here or it will explode due to the envvar `IN_NIX_SHELL` potentially not existing. I don't know a good way around this.